File: c++-srcs/combopt/mincov/gtest/MinCov_test.cc

/// @file MinCov_test.cc
/// @brief MinCov_test の実装ファイル
/// @author Yusuke Matsunaga (松永 裕介)
///
/// Copyright (C) 2026 Yusuke Matsunaga
/// All rights reserved.

#include <gtest/gtest.h>
#include "ym/MinCov.h"
#include "ym/McPerturbator.h"
#include "ym/JsonValue.h"
#include <random>


BEGIN_NAMESPACE_YM_MINCOV

TEST(MinCovTest, greedy)
{
  SizeType nr = 3;
  SizeType nc = 4;
  auto elem_list = std::vector<MinCov::ElemType>{
    {0, 0}, {0, 2},
    {1, 1}, {1, 2},
    {2, 0}, {2, 3}
  };

  MinCov mincov(nr, nc);
  for ( auto& elem: elem_list ) {
    mincov.insert_elem(elem.row_pos, elem.col_pos);
  }

  EXPECT_EQ( nr, mincov.row_size() );
  EXPECT_EQ( nc, mincov.col_size() );

  auto option = JsonValue::object();
  option.add("algorithm", JsonValue("greedy"));

  std::vector<SizeType> solution;
  auto ns = mincov.solve(solution, option);

  EXPECT_EQ( 2, ns );
}

TEST(MinCovTest, exact)
{
  SizeType nr = 3;
  SizeType nc = 4;
  auto elem_list = std::vector<MinCov::ElemType>{
    {0, 0}, {0, 2},
    {1, 1}, {1, 2},
    {2, 0}, {2, 3}
  };

  MinCov mincov(nr, nc);
  for ( auto& elem: elem_list ) {
    mincov.insert_elem(elem.row_pos, elem.col_pos);
  }

  EXPECT_EQ( nr, mincov.row_size() );
  EXPECT_EQ( nc, mincov.col_size() );

  auto option = JsonValue::object();
  option.add("algorithm", JsonValue("exact"));

  std::vector<SizeType> solution;
  auto ns = mincov.solve(solution, option);

  EXPECT_EQ( 2, ns );
}

TEST(MinCovTest, perturb)
{
  SizeType nr = 3;
  SizeType nc = 4;
  auto elem_list = std::vector<MinCov::ElemType>{
    {0, 0}, {0, 2},
    {1, 1}, {1, 2},
    {2, 0}, {2, 3}
  };

  MinCov mincov(nr, nc);
  for ( auto& elem: elem_list ) {
    mincov.insert_elem(elem.row_pos, elem.col_pos);
  }

  EXPECT_EQ( nr, mincov.row_size() );
  EXPECT_EQ( nc, mincov.col_size() );

  auto solution1 = std::vector<SizeType>{0, 1};

  ASSERT_TRUE( mincov.verify_solution(solution1) );

  McPerturbator perturb(mincov);

  auto solution2 = perturb(solution1);
  EXPECT_TRUE( mincov.verify_solution(solution2) );
}

END_NAMESPACE_YM_MINCOV

File: py_capi/scripts/combopt/mincov_test.py
#! /usr/bin/env python3

""" MinCov 用のスクリプト

:file: mincov_test.py
:author: Yusuke Matsunaga (松永 裕介)
:copyright: Copyright (C) 2026 Yusuke Matsunaga, All rights reserved.
"""

from argparse import ArgumentParser
import re
import time
from ymworks.combopt import MinCov
from ymworks.misc import JsonValue

pat1 = re.compile(r'(\d+)\s*(\d+)')
pat2 = re.compile(r'\s*-1\s*-1\s*')


def read_pi(filename):
    with open(filename, 'rt') as fin:
        elem_list = []
        max_row = 0
        max_col = 0
        for line in fin:
            if pat2.match(line):
                break;
            result = pat1.match(line)
            if not result:
                print(f'invalid line: {line}, ignored')
                continue
            row = int(result.group(1))
            col = int(result.group(2))
            elem_list.append((row, col))
        max_row = max(max_row, row)
        max_col = max(max_col, col)
        row_size = max_row + 1
        col_size = max_col + 1
        mincov = MinCov(row_size=row_size, col_size=col_size)
        for row, col in elem_list:
            mincov.insert_elem(row_pos=row, col_pos=col)
        return mincov
    assert False


parser = ArgumentParser()
parser.add_argument("filename", type=str)
parser.add_argument('--algorithm', type=str)

args = parser.parse_args()
if args is None:
    parser.usage()
    exit(1)

filename = args.filename
if args.algorithm is None:
    algorithm = "greedy"
else:
    algorithm = args.algorithm

s_time = time.time()
mincov = read_pi(filename)
e_time = time.time()
print(f'read_pi({filename}) end. Time: {e_time - s_time}')

option = JsonValue.object()
option.add_with_key('algorithm', JsonValue(algorithm))

s_time = time.time()
solution, nc = mincov.solve(option=option)
e_time = time.time()
print(f'solve end. Time: {e_time - s_time}')
